proposals/NNNN-metatype-keypath.md

```diff
@@ -0,0 +1,123 @@
+* Review Manager: TBD
+* Status: **Awaiting review**
+* Implementation: [apple/swift#73242](https://github.com/apple/swift/pull/73242)
+* Review: ([Pitch](https://forums.swift.org/t/pitch-metatype-keypaths/70767))
+
+## Introduction
+
+Key path expressions access properties dynamically. They are declared with a concrete root type and one or more key path components that define a path to a resulting value via the type’s properties, subscripts, optional-chaining expressions, forced unwrapped expressions, or self. This proposal expands key path expression access to include static properties of a type, i.e., metatype keypaths.
+
+## Motivation
+
+Metatype keypaths were briefly explored in the pitch for [SE-0254](https://forums.swift.org/t/pitch-static-and-class-subscripts/21850) and the [proposal](https://github.com/apple/swift-evolution/blob/main/proposals/0254-static-subscripts.md#metatype-key-paths) later recommended them as a future direction. Allowing key path expressions to directly refer to static properties has also been discussed on the Swift Forums for database lookups when used [in conjunction with @dynamicMemberLookup](https://forums.swift.org/t/dynamic-key-path-member-lookup-cannot-refer-to-static-member/30212) and as a way to avoid verbose hacks like [referring to a static property through another computed property](https://forums.swift.org/t/key-path-cannot-refer-to-static-member/28055). Supporting metatype keypaths in the Swift language will address these challenges and improve language semantics.
+
+## Proposed solution
+
+We propose to allow keypath expressions to define a reference to static properties. The following usage, which currently generates a compiler error, will be allowed as valid Swift code.
+
+```
+struct Bee {
+  static let name = "honeybee"
+}
+
+let kp = \Bee.Type.name
+```
+
+## Detailed design
+
+### Metatype syntax
+
+Keypath expressions where the first component refers to a static property will include `.Type` on their root types stated in the key path contextual type or in the key path literal. For example:
+
+```
+struct Bee {
+  static let name = "honeybee"
+}
+
+let kpWithContextualType: KeyPath<Bee.Type, String> = \.name // key path contextual root type of Bee.Type
+let kpWithLiteral = \Bee.Type.name // key path literal \Bee.Type
+```
+
+Attempting to write the above metatype keypath without including `.Type will trigger an error diagnostic:
+
+```
+let kpWithLiteral = \Bee.name // error: static member 'name' cannot be used on instance of type 'Bee'
+```
+
+Keypath expressions where the component referencing a static property is not the first component do not require `.Type`:
+```
+struct Species {
+  static let isNative = true
+}
+
+struct Wasp {
+  var species: Species.Type {Species.self}
+}
+
+let kpSecondComponentIsStatic = \Wasp.species.isNative
+```
+### Access semantics
+
+Immutable static properties will form the read-only keypaths just like immutable instance properties.
```

Comment:
Maybe it should be something like - Metatype key paths share almost all access semantics with instance member key paths with a notable exception to mutable static properties - such properties always form `ReferenceWritableKeyPath` because their base type (a metatype type) always has reference semantics.